### PR TITLE
Rework optimization of global variables

### PR DESCRIPTION
**Pull request by eeckstein**
The goal of this PR is to simplify the optimizations of global variables. 
The existing complex GlobalOpt optimization is replaced by individual small optimizations, which are implemented in swift:
* `InitializeStaticGlobals`: converts lazily initialized globals to a statically initialized global variables
* `ReadOnlyGlobalVariables`: marks global `var` variables as `let` if they are never written
* The simplification of `load` instructions replaces a load of a statically initialized global let with its initializer value
* The simplification of `builtin` instructions constant folds builtins and removes redundant `builtin "once"` calls

Those passes are very small and simple and in total much fewer lines of code than the old GlobalOpt pass.
Except the `ReadOnlyGlobalVariables`, those passes also run at Onone and are therefore important for performance annotated functions, which access global variables.

This PR also simplifies the inliners in the pass pipeline: three is no distinction between the "mid-level" and "late" inliner anymore. The only difference of the "mid-level" inliner compared to the "late" inliner was that it didn't line global accessors. This PR adds support for inlined global accessors in LICM and CSE (they now can handle `builtin "once"` calls). Therefore it's not needed to have both kinds of inliners.

The PR contains a few other small changes and fixes which are needed to make this change work. For details see the commit messages.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test windows

File: SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    if !g.isPossiblyUsedExternally,
+       !g.isLet,
+       !writtenGlobals.contains(g) {
+      g.setIsLet(to: true, moduleContext)
+    }
+  }
+}
+
+private extension Value {
+  var addressHasWrites: Bool {
```
Comment (atrick) [inline SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift:47]: It's very strange to me that you would hide the functionality of def-use traversal inside the unnecessary abstractions of a computed property within an extension of value. If something is initiating a def-use walk, I *really* want that to be obvious at the point-of-use. It's quite important that people don't accidentally perform repeated def-use walks.

File: SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+///
+let readOnlyGlobalVariablesPass = ModulePass(name: "read-only-global-variables") {
+  (moduleContext: ModulePassContext) in
+
+  var writtenGlobals = Set<GlobalVariable>()
+
+  for f in moduleContext.functions {
+    for inst in f.instructions {
+      if let gAddr = inst as? GlobalAddrInst {
+        if gAddr.addressHasWrites {
```
Comment (atrick) [inline SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift:30]: The most important part of this algorithm is that it performs a def-use traversal. Hiding the most important fact that code needs to convey is the definition of unreadable code.

**Review by atrick — COMMENTED** [review]
@eeckstein This is really great in terms of simplifying the compiler architecture, improving the optimization design, and eliminating confusing code on the C++ side.

The issues I see on implementation side are relatively minor now, but we need to consider setting precedent in new code base. It isn't obvious yet in this PR, but the use of extensions in the Swift sources to tack on arbitrary local functionality will gradually lead to bloated interfaces, incomprehensible layering, and widespread implicit `self` which makes the code unreadable. You don't need take my word for it, I'm just initiating the discussion you should have with others.

File: SwiftCompilerSources/Sources/Optimizer/PassManager/Context.swift [hunk 3]
```diff
@@ -71,6 +71,33 @@ extension MutatingContext {
     erase(instruction: inst)
   }
 
+  /// Copies all instructions of a static init value of a global to the insertion point of `builder`.
+  func copyStaticInitializer(fromInitValue: Value, to builder: Builder) -> Value? {
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply SwiftCompilerSources/Sources/Optimizer/PassManager/Context.swift:75]: In general, all SIL modifications must go through context, for two reasons:
* Automatic change notifications for analysis invalidation
* In future, it allows the context to be an "actor" in a concurrent pass pipeline. E.g. we can verify that only the current function is modified and not e.g. they body of a called function, etc.

Optimizer utilities can be either
* a member of the context, directly
* a free standing function, which gets the context and the SIL entity (e.g. the `Function`) as parameters
* a method in an extension of a SIL entity (e.g. `Function`) which gets the context as a parameter

I'm not sure what's the best way. But yes, it makes sense to agree on a "coding style" here.

**Comment by eeckstein** [comment]
@atrick thanks for reviewing!

> we need to consider setting precedent in new code base

I agree, we should agree on an "coding style".

> the use of extensions in the Swift sources to tack on arbitrary local functionality will gradually lead to bloated interfaces

Note that "local functionalities" are inside `private` extensions, which are only visible in that file. So it's not the case that such extensions will bloat the extended type's interface.

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/InitializeStaticGlobals.swift [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  /// instructions which initialize the global.
+  ///
+  /// The function's single basic block must contain following code pattern:
+  /// ```
+  ///   alloc_global @the_global
+  ///   %a = global_addr @the_global
+  ///   %i = some_const_initializer_insts
+  ///   store %i to %a
+  /// ```
+  func getGlobalInitialization() -> (allocInst: AllocGlobalInst, storeToGlobal: StoreInst)? {
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply SwiftCompilerSources/Sources/Optimizer/FunctionPasses/InitializeStaticGlobals.swift:76]: > As a rule, if it doesn't require access to Function's internal state, don't make it part of Function's interface.

Note that in an extension you can't access internal state and a private extension doesn't add to the type's interface.

File: SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift (hunk 1 above)
Comment (eeckstein) [reply SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift:47]: Well, a computed property is not more of an abstraction than a method.
But I agree that a property conveys that it's something very simple and doesn't involve any (potential) computational expensive operations.
So in this case it's maybe better to define it as a method.

**Comment by eeckstein** [comment]
Alternative: https://github.com/apple/swift/pull/65821

File: SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift (hunk 1 above)
Comment (atrick) [reply SwiftCompilerSources/Sources/Optimizer/ModulePasses/ReadOnlyGlobalVariables.swift:47]: > Well, a computed property is not more of an abstraction than a method.
But I agree that a property conveys that it's something very simple (O(1) complexity) and doesn't involve any (potential) computational expensive operations.

@eeckstein Yes, you said it better. That's the only point I was making. I've seen the same argument made during review of Swift libraries.

**Pull request merged**